### ISSUE TITLE
False positive unreachable for generator that might not yield

### ISSUE THREAD

**Issue by AckslD** [python/mypy#12587]
**Bug Report**
(Maybe related to #8766)

Consider the following code:
```python
from typing import Iterator


def gen() -> Iterator[int]:
    if 1 == 2:
        yield 0


def test() -> None:
    x = None
    for x in gen():
        pass
    if x is None:
        print('reachable')
```

**To Reproduce**
```
mypy above_file.py --warn-unreachable
```

**Expected Behavior**
No error.

**Actual Behavior**
```
tmp.py:14: error: Statement is unreachable
            print('reachable')
            ^
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.931
- Mypy command-line flags: --warn-unreachable
- Python version used: Python 3.9.10
- Operating system and version: arch 5.17.1-arch1-1

**Comment by JelleZijlstra** [comment]
I think there's an existing issue for this: the problem is essentially that mypy thinks for loops always execute at least once.

**Comment by AlexWaygood** [comment]
> I think there's an existing issue for this

#8637?

**Comment by AckslD** [comment]
Thanks both!
